homeassistant/components/netgear/device_tracker.py

```diff
@@ -1,166 +1,141 @@
-            if tracked:
-                devices.append(dev.mac)
-                if (
-                    self.tracked_accesspoints
-                    and dev.conn_ap_mac in self.tracked_accesspoints
-                ):
-                    devices.append(f"{dev.mac}_{dev.conn_ap_mac}")
-
-        return devices
-
-    def get_device_name(self, device):
-        """Return the name of the given device or the MAC if we don't know."""
-        parts = device.split("_")
-        mac = parts[0]
-        ap_mac = None
-        if len(parts) > 1:
-            ap_mac = parts[1]
-
-        name = None
-        for dev in self.last_results:
-            if dev.mac == mac:
-                name = dev.name
-                break
-
-        if not name or name == "--":
-            name = mac
+from homeassistant.components.device_tracker import SOURCE_TYPE_ROUTER
+from homeassistant.components.device_tracker.config_entry import ScannerEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import callback
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import HomeAssistantType
 
-        if ap_mac:
-            ap_name = "Router"
-            for dev in self.last_results:
-                if dev.mac == ap_mac:
-                    ap_name = dev.name
-                    break
+from .const import DEVICE_ICONS, DOMAIN
+from .router import NetgearRouter
 
-            return f"{name} on {ap_name}"
-
-        return name
-
-    def _update_info(self):
-        """Retrieve latest information from the Netgear router.
-
-        Returns boolean if scanning successful.
-        """
-        _LOGGER.debug("Scanning")
-
-        results = self.get_attached_devices()
-
-        if _LOGGER.isEnabledFor(logging.DEBUG):
-            _LOGGER.debug("Scan result: \n%s", pformat(results))
-
-        if results is None:
-            _LOGGER.warning("Error scanning devices")
+_LOGGER = logging.getLogger(__name__)
 
-        self.last_results = results or []
 
-    def get_attached_devices(self):
-        """List attached devices with pynetgear.
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up device tracker for Netgear component."""
+    router = hass.data[DOMAIN][entry.unique_id]
+    tracked = set()
+
+    @callback
+    def update_router():
+        """Update the values of the router."""
+        add_entities(router, async_add_entities, tracked)
+
+    router.listeners.append(
+        async_dispatcher_connect(hass, router.signal_device_new, update_router)
+    )
+
+    update_router()
+
+
+@callback
+def add_entities(router, async_add_entities, tracked):
+    """Add new tracker entities from the router."""
+    new_tracked = []
+
+    for mac, device in router.devices.items():
+        if mac in tracked:
+            continue
```

Comment:
Instead of this, just always add all entities and then let the Home Assistant UI deal with disabling the unwanted entities.

Reply:
This is not to don't track a device, but to add a new `NetgearDevice(ScannerEntity)` when it's been connected/discovered to the router after HA start.

Did the same in Freebox, which use Unifi as a model.

Reply:
Unifi allows specifying groups of devices to be tracked: wired and wireless. Individual disabling of entities should happen via the Home Assistant UI.

Reply:
Can you re think this comment ?

Because this is not due to disable entries or not.

With the `CONF_DEVICES` and `CONF_EXCLUDE` the router.py did.

---
Here is what happens:

- HA boot
- configuration setup
- platform setup
  - async_setup_entry: add listener to add devices in the future (see later)
  - async_setup_entry: call `update_router` that add current connected devices with `add_entities`
  - async_setup_entry: pass `tracked` var that track all devices added to HA (set of mac address)
  - add_entities: the `new_tracked` is the list of connected devices that are not yet added to HA
  - add_entities: compare if current connected device is already added to HA `if mac in tracked`, if not, add to `new_tracked` and `tracked`
  - add_entities: add devices to HA for all `new_tracked` devices
- When a new device is connected to the router, the router call `signal_device_new`, which call `update_router` and add only the new device.

Reply:
Yeah, this is just so that we don't create duplicate entities for the same mac address. Each mac address should be one entity.